lib/reek/tree_walker.rb

```diff
@@ -105,17 +111,19 @@ def process_default(exp)
 
     def process_args(_) end
 
-    # :reek:TooManyStatements: { max_statements: 6 }
+    # :reek:TooManyStatements: { max_statements: 8 }
     # :reek:FeatureEnvy
     def process_send(exp)
+      method_name = exp.method_name
       if exp.visibility_modifier?
-        element.track_visibility(exp.method_name, exp.arg_names)
-      end
-      if exp.attribute_writer?
+        element.track_visibility(method_name, exp.arg_names)
+      elsif exp.attribute_writer?
         exp.args.each do |arg|
           next unless arg.type == :sym
           new_context(Context::AttributeContext, arg, exp)
         end
+      else
+        new_context(Context::SendContext, exp, method_name)
```

Comment:
I understand your reasoning, but I believe we should consider Contexts to be 'things that can have a smell', so currently these would be classes, modules, methods, and some class-level method calls like attribute definitions.

In contrast, all other additions to Nodes are in TreeDresser.

We could merge contexts and sexp extensions, but I think it makes sense to have two hierarchies, since the nodes often have intermediate hierarchy levels that we don't want to include in the contexts. This was one reason I moved the visibility tracking into contexts.

What do you think?